zerver/lib/push_notifications.py

```diff
@@ -7,27 +7,106 @@
 from zerver.lib.timestamp import timestamp_to_datetime
 from zerver.decorator import statsd_increment
 
-from apnsclient import Session, Message, APNs
-from apnsclient.apns import Connection
+from apns import APNs, Frame, Payload, SENT_BUFFER_QTY
 import gcmclient
 
 from django.conf import settings
 
-import base64, binascii, logging, os
+import base64, binascii, logging, os, random, time
+from collections import OrderedDict
+from functools import partial
+
+class FifoDict(OrderedDict):
+    def __init__(self, *args, **kwargs):
+        self.size = kwargs.pop('size', 'None')
+        super(FifoDict, self).__init__(*args, **kwargs)
+
+    def __setitem__(self, key, value):
+        super(FifoDict, self).__setitem__(key, value)
+        if len(self) > self.size:
+            self.popitem(last=False)
+
+# APNS error codes
+ERROR_CODES = {
+    1: 'Processing error',
+    2: 'Missing device token', # looks like token was empty?
+    3: 'Missing topic', # topic is encoded in the certificate, looks like certificate is wrong. bail out.
+    4: 'Missing payload', # bail out, our message looks like empty
+    5: 'Invalid token size', # current token has wrong size, skip it and retry
+    6: 'Invalid topic size', # can not happen, we do not send topic, it is part of certificate. bail out.
+    7: 'Invalid payload size', # our payload is probably too big. bail out.
+    8: 'Invalid token', # our device token is broken, skipt it and retry
+    10: 'Shutdown', # server went into maintenance mode. reported token is the last success, skip it and retry.
+    None: 'Unknown', # unknown error, for sure we try again, but user should limit number of retries
+}
```

Comment:
Nope. These are the error code returned by APNS servers. They are specified in APNS documentation, however, I copied them from the old library, `apnsclient`, that we were using.